File: backend/app/routers/tracker.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from app.models.tracker import NewTrackerRow, TrackerRow
from app.services import tracker_db as tracker_sheets

router = APIRouter(prefix="/tracker", tags=["tracker"])


@router.get("/rows", response_model=list[TrackerRow])
def list_rows():
    try:
        return tracker_sheets.get_all_rows()
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Could not fetch tracker rows: {exc}")


@router.post("/rows", status_code=201)
def add_row(payload: NewTrackerRow):
    try:
        tracker_sheets.append_row(
            date=payload.date,
            channel_name=payload.channel_name,
            views=payload.views,
            minutes_watched=payload.minutes_watched,
        )
        return {"message": "Row added"}
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Could not add row: {exc}")


@router.delete("/rows/{row_index}", status_code=204)
def remove_row(row_index: int):
    try:
        tracker_sheets.delete_row(row_index)
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Could not delete row: {exc}")


class NewChannel(BaseModel):
    name: str


@router.get("/channels", response_model=list[str])
def list_channels():
    try:
        return tracker_sheets.get_channels()
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Could not fetch channels: {exc}")


@router.post("/channels", status_code=201)
def create_channel(payload: NewChannel):
    name = payload.name.strip()
    if not name:
        raise HTTPException(status_code=422, detail="Channel name cannot be empty")
    try:
        tracker_sheets.add_channel(name)
        return {"name": name}
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Could not create channel: {exc}")

File: frontend/src/api/tracker.ts
import { api } from "./client";

export interface TrackerRow {
  row_index: number;
  date: string;
  channel_name: string;
  views: number;
  minutes_watched: number;
}

export interface NewTrackerRow {
  date: string;
  channel_name: string;
  views: number;
  minutes_watched: number;
}

export async function getTrackerRows(): Promise<TrackerRow[]> {
  const { data } = await api.get<TrackerRow[]>("/tracker/rows");
  return data;
}

export async function addTrackerRow(row: NewTrackerRow): Promise<void> {
  await api.post("/tracker/rows", row);
}

export async function deleteTrackerRow(rowIndex: number): Promise<void> {
  await api.delete(`/tracker/rows/${rowIndex}`);
}

export async function getChannels(): Promise<string[]> {
  const { data } = await api.get<string[]>("/tracker/channels");
  return data;
}

export async function addChannel(name: string): Promise<void> {
  await api.post("/tracker/channels", { name });
}
